lib/puppet/file_system/windows.rb

```diff
@@ -114,8 +118,136 @@ def read_preserve_line_endings(path)
+             dacl
+           when 0640, 0600 # Setting both of these with Group Read Access
+             dacl = secure_dacl(current_sid, FILE_RW, FILE_READ)
+             dacl
+           when 0660
+             dacl = secure_dacl(current_sid, FILE_RW, FILE_RW)
+             dacl
+           when 0664
+             dacl = secure_dacl(current_sid, FILE_RW, FILE_RW)
+             dacl.allow(Puppet::Util::Windows::SID::BuiltinUsers, FILE_READ)
+             dacl.allow(Puppet::Util::Windows::SID::Everyone, FILE_READ)
+             dacl
+           when 0666
+             dacl = secure_dacl(current_sid, FILE_RW, FILE_RW)
+             dacl.allow(Puppet::Util::Windows::SID::BuiltinUsers, FILE_RW)
+             dacl.allow(Puppet::Util::Windows::SID::Everyone, FILE_RW)
+             dacl
+           when 0444
+             dacl = secure_dacl(current_sid, FILE_READ, FILE_READ)
+             dacl.allow(Puppet::Util::Windows::SID::BuiltinUsers, FILE_READ)
+             dacl.allow(Puppet::Util::Windows::SID::Everyone, FILE_READ)
+             dacl
+           when 0440
+             dacl = secure_dacl(current_sid, FILE_READ, FILE_READ)
+             dacl
+            when nil
+             get_dacl_from_file(path) || secure_dacl(current_sid)
+           else
+             raise ArgumentError, "#{mode} is invalid: Only modes 0644, 0640, 0660, 0666, 0600 and 0440 are allowed"
+           end
+
+    # puts "replace_file: dacl: #{dacl.inspect}"
+    tempfile = Puppet::FileSystem::Uniquefile.new(Puppet::FileSystem.basename_string(path), Puppet::FileSystem.dir_string(path))
+    begin
+      tempdacl = Puppet::Util::Windows::AccessControlList.new
+      tempdacl.allow(current_sid, FULL_CONTROL)
+      set_dacl(tempfile.path, tempdacl)
+      # puts "replace_file: tempfile 1 work"
+
+      begin
+        yield tempfile
+        tempfile.flush
+        tempfile.fsync
+      ensure
+        tempfile.close
+      end
+
+      set_dacl(tempfile.path, dacl) if dacl
+      File.rename(tempfile.path, Puppet::FileSystem.path_string(path))
+    ensure
+      tempfile.close!
+    end
+  rescue Puppet::Util::Windows::Error => e
+    case e.code
+    when ACCESS_DENIED, SHARING_VIOLATION, LOCK_VIOLATION
+      raise Errno::EACCES.new(Puppet::FileSystem.path_string(path), e)
+    else
+      raise SystemCallError.new(e.message)
+    end
+  end
+
   private
 
+  def set_dacl(path, dacl)
+    # Set the DACL
+    # This has a "special case" - if both Owner and Group are SYSTEM, then the Group field
+    # is set to BuiltinAdministrators (PUP-9719). This is needed to ensure that if Puppet Agent
+    # is run under SYSTEM/SYSTEM (e.g. under AWS services), that all internally managed Puppet files
+    # are are still writeable from the Administrator account
+    sd = Puppet::Util::Windows::Security.get_security_descriptor(path)
+    sd_group = sd.group == sd.owner && sd.group == Puppet::Util::Windows::SID::LocalSystem ? Puppet::Util::Windows::SID::BuiltinAdministrators : sd.group
```

Comment:
This line is too long and it's calling sd.group at least 2 times.
It could be written like this:
```ruby
sd_group = sd.group
sd_owner = sd.owner

if sd_group == sd_owner && sd_group == Puppet::Util::Windows::SID::LocalSystem
  sd_group = Puppet::Util::Windows::SID::BuiltinAdministrators
end

new_sd = Puppet::Util::Windows::SecurityDescriptor.new(sd_owner, sd_group, dacl, true)
```